### PR TITLE
Fixes #26729 - add polling to API middleware

### PR DESCRIPTION
**Pull request by Ron-Lavi**
current API action looks like this:
```
const some_action = payload => ({
  type: API_OPERATIONS.GET,
  key: MY_API_REQUEST_SPECIAL_KEY,
  payload,
});
```
with polling:
```
const some_action = payload => ({
  type: API_OPERATIONS.GET,
  key: MY_API_REQUEST_SPECIAL_KEY,
  polling: 3000, /** value in ms, or true which will use the default values. */  
  payload,
});
```
How to stop the polling?
_index.js_
```
import { APIActions } from "../../redux/API";
...
const mapDispatchToProps = dispatch => bindActionCreators( { ...actions, ...APIActions }, dispatch);
```
and then use it in your component:
```
handlePolling = () => {
  this.props.stopPolling(MY_API_REQUEST_SPECIAL_KEY)
}
```

### CODE REVIEW

**Comment by sharvit** [comment]
Nice work @laviro :+1: 

Woudn't it make more sense to first implement the base API reducer? (So this PR can add more functionality based on it)

**Comment by Ron-Lavi** [comment]
Thanks @sharvit ! added more changes,
and btw  @amirfefer I tested it with multiple polling processes and it works well,
we could add the reducer in a separate PR though I prefer not to create an empty reducer,
In this PR we are introducing two new cases, and in the following PRs we can amend it :)

**Comment by Ron-Lavi** [comment]
Blocked by #6739

**Comment by MariaAga** [comment]
I tested it and it works great in foreman tasks, can you add the info about polling to the documentation? `webpack/stories/docs/APIMiddleware.md`

**Comment by ezr-ondrej** [comment]
@laviro could you please rebase, #6739 is merged.

**Comment by Ron-Lavi** [comment]
Thanks @ezr-ondrej , rebased.
@MariaAga I added the readme file to the storybook :)

**Comment by ohadlevy** [comment]
Can we use it already? E.g the notification drawer?

**Comment by Ron-Lavi** [comment]
@ohadlevy yep, that's my follow up PR :)

File: webpack/stories/docs/APIPollingMiddleware.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  key: MY_API_REQUEST_SPECIAL_KEY,
+  polling: 3000, /** value in ms, or true which will use the default values. */  
+  payload,
+});
+```
+
+To stop the polling, we will need to use the "stopPolling" Action from APIActions:
+```js
+// MyComponent/index.js
+import { APIActions } from "../../redux/API";
```
Comment (MariaAga) [inline webpack/stories/docs/APIPollingMiddleware.md:27]: Please add a `foremanReact` route for plugins

**Review by MariaAga — DISMISSED** [review]
LGTM, just missing a line in the docs :)

**Review by amirfefer — COMMENTED** [review]
@laviro , according to @ohadlevy IMHO it makes sense to add at least one use for this middleware in this PR -  sometimes we figure out that some changes are required to the initial implementation, plus it won't make this PR much bigger (even would reduce some amount of code due to the reuse of redux) and a real example would be great so other plugins can use it asap :)

**Comment by Ron-Lavi** [comment]
@ohadlevy @amirfefer Added a follow up PR in #7121
so we could also continue the relevant discussion over there

**Comment by Ron-Lavi** [comment]
just rebased, and fixed the storybook documentation @MariaAga

**Comment by Ron-Lavi** [comment]
Updated with testing edge-cases where someone tries to register an existing polling `key`,
or remove a `key` which doesn't exist.
Added tests too.

**Review by MariaAga — DISMISSED** [review]

File: webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
+    const APIRequest = () =>
+      get(payload, url, store, actionTypeGenerator(key, actionTypes));
+    if (polling) {
+      const isAlreadyPolling = selectPollingID(store.getState(), key);
+      if (isAlreadyPolling) {
+        throw registeredPollingException(key);
+      }
+      next(startPolling(key, APIRequest, polling));
+    } else {
+      APIRequest();
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js]: don't we need to use `next` here?
AFAIK if we will have another  middleware it won't be chained and it will be lost

**Review by amirfefer — COMMENTED** [review]
Thanks @laviro ! Great middlewere :+1: 
Some thoughts:
1. in order to stop the polling we can use redux-hooks instead of using `mapsDispatchToProps`
so the consumer doesn't need to use the expansive `connect`, instead the consumer can use "useDipacth"

2. how about testing the middlewere itself, including verifying that `next`works as expected
[this blog](http://jonnyreeves.co.uk/2016/redux-middleware/) shows a nice flow of testing a middlewere

**Comment by Ron-Lavi** [comment]
Thank you @amirfefer, great catch ! :dart: :angel: 
according to [this answer](https://stackoverflow.com/a/45964310) we should always return next(action)
to chain middlewares. it should be documented better on the redux website I guess.

@MariaAga This cause snapshots changes in other tests since we are also returning the original action, could you take a look?

@amirfefer I will investigate more about the redux hook, so we can show another example of how to use the `stopPolling` action.
and I will also add more tests for the middleware itself, thanks !

**Comment by MariaAga** [comment]
@laviro Since you change the tests to match snapshots can you remove the fixtures from the fixtures files?

**Comment by Ron-Lavi** [comment]
@MariaAga eventually I updated the fixtures, it reduced lines of code by 400 !

File: webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.fixtures.js [hunk 3]
```diff
@@ -23,52 +23,87 @@ export const successRequestData = [
...
-  },
-  {
-    payload: {
-      data: [['centOS 7.1', 6]],
-      ...payloads.operatingsystem,
+const actions = {
+  architectures: {
+    REQUEST: {
+      payload: {
+        ...payloads.architecture,
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.fixtures.js]: Can this be just 
`payload : payloads.architecture` ?

**Review by MariaAga — COMMENTED** [review]
The snapshots are good now with the `action(next)` update, thanks @laviro

**Comment by Ron-Lavi** [comment]
@MariaAga, @amirfefer resolved everything that you suggested, thanks :)

File: webpack/assets/javascripts/react_app/redux/API/APIReducer.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+const { START_POLLING, STOP_POLLING } = API_OPERATIONS;
+const initialState = Immutable({
+  polling: {},
+});
+
+export const reducer = (state = initialState, action) => {
+  const { type, key, payload: { APIRequest, polling } = {} } = action;
+  switch (type) {
+    case START_POLLING:
+      const pollingID = startAPIInterval(APIRequest, polling);
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIReducer.js]: I think this kind of logic shouldn't be in a reducer, it dispatches another action. which triggers another reducer.
Reducers should just store the new state, and I highly recommend to move this to the middleware itself and dispatch actions there

File: webpack/assets/javascripts/react_app/redux/API/APIReducer.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        ...state.polling,
+        [key]: pollingID,
+      });
+    case STOP_POLLING:
+      const pollingProcessID = state.polling[key];
+      if (!pollingProcessID) {
+        console.error(unregisteredPollingException(key));
+        return state;
+      }
+      stopAPIInterval(pollingProcessID);
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIReducer.js]: ditto

File: webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js [hunk 6]
```diff
@@ -1,11 +1,29 @@
...
+      if (isAlreadyPolling) {
+        console.error(registeredPollingException(key));
+        return next(action);
+      }
+      /** 
+        force the APIRequest for the first time,
+        instead of waiting for the polling to start.
+      */
+      APIRequest();
+      return next(startPolling(key, APIRequest, polling));
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js]: After some investigation, triggering `next(<another action>)` will not dispatch it from the beginning of the middlewares stack, so in this case, we need to use `dispatch`, which will run all the middlewares

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @laviro ! 
Does it work with plugins out of the box?
 would it make sense to add a plugin example in the docs?

File: webpack/assets/javascripts/react_app/redux/API/APIReducer.js (hunk 4 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIReducer.js]: `startApiInterval` is not an action and does not trigger another reducer,
it is just using `window.setInterval` method, but I could move it to the middleware itself or to the action.

File: webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js (hunk 6 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js]: I couldn't find information about such an issue nor in the Redux middleware documentation,
could you explain more? from what I found returning `next(action)` worked as expected.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js]: after investigating more, and we talked about it offline, I removed the return and next
and instead use dispatch from the startPolling action. thanks @amirfefer

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js]: `next(action)` will continue this action to the next middleware in the chain.
but prior middlewares won't be triggered, only `diaptch` will trigger all of the chain, so if it is a new action (like `startPolling`) and not the given action, some middlewares could be missed.

File: webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js [hunk 7]
```diff
@@ -1,11 +1,17 @@
+/* eslint-disable no-console */
 import { API_OPERATIONS, actionTypeGenerator } from './';
 import { get } from './APIRequest';
+import { startPolling } from './APIActions';
 
 export const APIMiddleware = store => next => action => {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js:5]: How about making a new `Interval` middleware, so the APIMiddleware will handle the requests and the interval will trigger an Action interval - with that we can make an interval not only for GET request, but for every action.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js:5]: I realy like the idea and I will investigate more about it,
What other intervals do we need?
Is it possible that using more middlewars will slow down performance as they are getting called by every action?

Comment (sharvit) [reply webpack/assets/javascripts/react_app/redux/API/APIMiddleware.js:5]: > What other intervals do we need?

Can't think about any but IMO if we can make it more reusable and independent let's go for it :+1: 

> Is it possible that using more middlewars will slow down performance as they are getting called by every action?

IMO it should not affect performance at all. (if the condition will be simple as it should)

**Comment by Ron-Lavi** [comment]
Updated with the new intervals middleware,
still need to add some tests and fix the readme.

**Comment by Ron-Lavi** [comment]
oops, closed by mistake and couldn't re-open,
Moved to a new PR: https://github.com/theforeman/foreman/pull/7157
